ide/posts/2020-06-50-ghcide-and-nixpkgs.md

```diff
@@ -0,0 +1,321 @@
+code. If you just want to know how to use ghcide in a nix environment jump to
+the end of the post or just consult the README of ghcide.
+
+<!--more-->
+
+## What’s the problem?
+
+Haskell development tooling setup is infamous for being brittle and hard to
+setup. Every other day when someone asks on reddit or in the `#haskell`
+channel, inescapably there will come at least one answer of the form "It’s not
+worth the pain. Just use ghcid." I guess one point of this blog series is that
+this does not have to be the case anymore.
+
+But what were the reasons for this resignation? One is certainly that `ghcid`
+is a really great and easy to use tool. But I think it‘s clear that a well done
+language server can leverage you much further and to me `ghcide` has already
+proven this.
+
+### Compile your project and ghcide with the same ghc!
+
+One source of frustration is likely that succesfully setting up a language
+server that is deeply interwoven with `ghc` like `ghcide` has one very
+important requirement. **You need to compile `ghcide` with the same `ghc`
+(version) as your project.** This shouldn‘t be hard to achieve nowadays - I’ll
+show how to do it if you use `nix` in this blogpost and I assume it‘s the
+default in other setups - but if you fail to meet this requirement you are in
+for a lot of trouble.
+
+So why exactly do we need to use "the same ghc" and what does that even mean?
+Frankly I am not totally sure. I am not a `ghcide` developer. I guess sometimes
+you can get away with some slight deviations. But the general recommendation is
+to use the same ghc version. I can tell you three situations that will cause
+problems or have caused problems for me:
+
+1. Using another `ghc` release. E.g. using `ghcide` compiled with `ghc` 8.6 on
+   a `ghc` 8.8 project.
+2. Using the same `ghc` release but with a patch to `ghc`. This happened to me
+   while using the `obelisk` framework which uses a modified `ghc`.
+3. Using the same `ghc` version but linked against different external libraries
+   like `glibc`. This can happen when different releases of nixpkgs are
+   involved.
+
+To sum up, both ghcs should come from the same source and be linked against the
+same libraries. Your best bet is to use the same binary. But that is not
+necessary.
+
+### How to actually get a ghcide on nix
+
+So consider the following situation. You are using nix to develop your Haskell
+project. To be more specific: To compile your project you use ghc installed
+from nixpkgs. This probably means that you have a `default.nix` and a
+`shell.nix` file in your project.
+
+So what are your ways to get a ghcide?
+
+1. Use the `ghcide-nix` derivation.
+2. Install `haskellPackages.ghcide` from nixpkgs.
+3. Compile one yourself using cabal.
+
+So I was developing an obelisk project and wanted to use ghcide. The first way
+was the recommend one in the `ghcide` README, but sadly that didn‘t work
+because (as I learned later from Elliot Cameron) the obelisk project uses a
```

Comment:
I think I wasn't the one who told you this because I didn't know either at the time!